[Feature][Transform-V2] Support multimodal embeddings

xiaochen-zhou opened the pull request:
<!--

Thank you for contributing to SeaTunnel! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

## Contribution Checklist
  - Make sure that the pull request corresponds to a [GITHUB issue](https://github.com/apache/seatunnel/issues).
  - Name the pull request in the form "[Feature] [component] Title of the pull request", where *Feature* can be replaced by `Hotfix`, `Bug`, etc.
  - Minor fixes should be named following this pattern: `[hotfix] [docs] Fix typo in README.md doc`.
-->

### Purpose of this pull request

**Multimodal Vectorization:**
```hocon
vectorization_fields {
      description_vector = description

      product_image_vector = {
        field = product_image_url
        modality = jpeg
        format = url
      }
      
      thumbnail_vector = {
        field = thumbnail_image
        modality = png
        format = binary
      }
 }

```

### Does this PR introduce _any_ user-facing change?

Yes


### How was this patch tested?

Add new tests


### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)

xiaochen-zhou committed:
[Feature][TransForm-V2] Support multimodal embeddings

update tests

xiaochen-zhou committed:
Merge branch 'dev' into support-multimodal

# Conflicts:
#	seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/nlpmodel/embedding/EmbeddingTransform.java
#	seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/nlpmodel/embedding/remote/doubao/DoubaoModel.java

corgy-w commented on docs/en/transform-v2/embedding.md:
```diff
@@ -19,7 +19,7 @@ different API endpoints.
+| vectorization_fields           | map    | yes      | -             | A mapping between input fields and their corresponding output vector fields. Supports multimodal field specifications with format `field_name:modality_type`.        |
 | model                          | string | yes      | -             | The specific model to use for embedding (e.g: `text-embedding-3-small` for OPENAI).                                                                                     |
 | api_path                       | string | no       | -             | The API endpoint for the embedding service. Typically provided by the model provider.                                                                                   |
 | dimension                      | int    | no       | -             | TThe vector dimension defaults to 2048. The Embedding-3 model supports custom vector dimensions, and it is recommended to choose dimensions of 256, 512, 1024, or 2048. |
```
Please help modify it. There's an extra "T" here

xiaochen-zhou replied:
> Please help modify it. There's an extra "T" here

OK.

corgy-w commented on seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/nlpmodel/embedding/multimodal/MultimodalField.java:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        String trimmedSpec = fieldSpec.trim();
+        if (trimmedSpec.contains(":")) {
```
When creating columns in the database, it is possible to add `:`, but is it incompatible here

xiaochen-zhou replied:
> When creating columns in the database, it is possible to add `:`, but is it incompatible here

```
vectorization_fields {
    text_vector {
        field = "text_field"
        modality = "text"
        format= "binary"

    }
    image_vector {
        field = "image_data"
        modality = "image"
        format= "url"
    }
} 
```

To ensure compatibility in this situation, what do you think about this format?

Hisoka-X commented:
1. How to get the url base on seatunnel workflow?
2. we have binary format in file source, I think we can direct read video or image from file then use base64 send to embedding provider. https://www.volcengine.com/docs/82379/1409291#base64-%E7%BC%96%E7%A0%81%E8%BE%93%E5%85%A5

xiaochen-zhou commented:
> 1. How to get the url base on seatunnel workflow?
> 2. we have binary format in file source, I think we can direct read video or image from file then use base64 send to embedding provider. https://www.volcengine.com/docs/82379/1409291#base64-%E7%BC%96%E7%A0%81%E8%BE%93%E5%85%A5

OK.

xiaochen-zhou commented:
> 1. How to get the url base on seatunnel workflow?
> 2. we have binary format in file source, I think we can direct read video or image from file then use base64 send to embedding provider. https://www.volcengine.com/docs/82379/1409291#base64-%E7%BC%96%E7%A0%81%E8%BE%93%E5%85%A5

Sometimes users may directly store image or video paths in the columns. Should we consider both scenarios, or just focus on one of them?

```
vectorization_fields {
    text_vector {
        field = "text_field"
        modality = "text"
        format= "binary"

    }
    image_vector {
        field = "image_data"
        modality = "image"
        format= "url"
    }
} 
```

Hisoka-X commented:
> > 1. How to get the url base on seatunnel workflow?
> > 2. we have binary format in file source, I think we can direct read video or image from file then use base64 send to embedding provider. https://www.volcengine.com/docs/82379/1409291#base64-%E7%BC%96%E7%A0%81%E8%BE%93%E5%85%A5
> 
> Sometimes users may directly store image or video paths in the columns. Should we consider both scenarios, or just focus on one of them?
> 
> ```
> vectorization_fields {
>     text_vector {
>         field = "text_field"
>         modality = "text"
>         format= "binary"
> 
>     }
>     image_vector {
>         field = "image_data"
>         modality = "image"
>         format= "url"
>     }
> } 
> ```

both.

xiaochen-zhou committed:
update code

1

1

1

1

1

update doc

[Improve][Csv] support configurable CSV delimiter in file connector (#9660)

[Fix][Doc] Fix rest api finished-jobs doc miss `SAVEPOINT_DONE` (#9676)

[Fix] [connector-jdbc] prevent precision loss in Float to BigDecimal conversion (#9670)

[Feature][Connector-File-Hadoop]Support multi table sink feature for HdfsFile (#9651)

[HotFix][CI] The Some CI subtasks are not executed (#9684)

[Docs][Connector-Iceberg] Update Iceberg doc with support S3 Tables rest catalogs (#9686)

[Docs] update jdbc related document (#9679)

[Fix][Connector-V2] Fix misleading parameter name in DorisStreamLoad (#9685)

[Feature] [connector-file] Add configurable sheet_max_rows support for Excel sink connector (#9668)

[fix][connectors-v2] repeated commit cause task exceptions (#9665)

[Feature][connector-v2] Add Sensorsdata Connector Support #9323 (#9432)

[Improve][API] Optimize the enumerator API semantics and reduce lock calls at the connector level (#9671)

[Hotfix][Zeta] Fix custom https port configuration not working (#9705)

[Improve][Doc] Add SeaTunnel tools into Readme (#9707)

[Improve][Core] Unify the aws-sdk-v2 version to 2.31.30 (#9698)

[Feature][Transform-V2] Add `TRIM_SCALE` function for sql transform (#9700)

[Improve][Core] Update apache common to apache common lang3 (#9694)

Co-authored-by: litiliu <litiliu@cisco.com>

xiaochen-zhou committed:
Merge branch 'dev' into support-multimodal

xiaochen-zhou commented:
@Hisoka-X @corgy-w Done.

Hisoka-X commented on seatunnel-transforms-v2/src/main/java/org/apache/seatunnel/transform/nlpmodel/embedding/EmbeddingTransform.java:
```diff
@@ -180,33 +187,49 @@ public void open() {
+            Object[] fieldValues = new Object[fieldOriginalIndexes.size()];
+            List<ByteBuffer> vectorization;
+            int i = 0;
+            for (Integer fieldOriginalIndex : fieldOriginalIndexes) {
+                FieldSpec fieldSpec = fieldSpecMap.get(fieldOriginalIndex);
+                Object value = inputRow.getField(fieldOriginalIndex);
+                fieldValues[i++] =
+                        isMultimodalFields ? new MultimodalFieldValue(fieldSpec, value) : value;
```
In the binary format scenario, the data contained in each row may be part of the complete file. We need to cache it in the transform and perform embedding after obtaining the complete file. Please refer to https://github.com/apache/seatunnel/blob/dev/seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/reader/BinaryReadStrategy.java#L132 and https://github.com/apache/seatunnel/blob/dev/seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/sink/writer/BinaryWriteStrategy.java#L78

Hisoka-X reviewed (commented):
Thanks @xiaochen-zhou . Overall looking great, only one problem need be solved.

xiaochen-zhou committed:
support binary chunk embedding

xiaochen-zhou committed:
fix test failure

xiaochen-zhou committed:
update e2e test

xiaochen-zhou committed:
Merge branch 'dev' into support-multimodal

xiaochen-zhou committed:
fix e2e failure

[Improve][CI] Disable close stable issue and JdbcVerticaIT (#9742)

[Improve][Doc] Update sql function result type (apache#9703) (#9711)

Hisoka-X commented on seatunnel-e2e/seatunnel-transforms-v2-e2e/seatunnel-transforms-v2-e2e-part-1/src/test/resources/embedding_transform_binary.conf:
```diff
@@ -0,0 +1,88 @@
+    file_format_type = "binary"
+    binary_complete_file_mode = false
+    binary_chunk_size = 1024
+    schema {
+      table = "cat"
+    }
```
what's for?

xiaochen-zhou committed:
Merge branch 'dev' into support-multimodal

xiaochen-zhou committed:
fix e2e failure

xiaochen-zhou committed:
Merge branch 'dev' into support-multimodal

# Conflicts:
#	seatunnel-engine/seatunnel-engine-server/src/test/java/org/apache/seatunnel/engine/server/rest/RestApiHttpsForTruststoreTest.java

xiaochen-zhou committed:
fix e2e failure

xiaochen-zhou committed:
fix e2e failure

Hisoka-X commented on seatunnel-api/src/main/java/org/apache/seatunnel/api/table/type/SeaTunnelRow.java:
```diff
@@ -126,6 +126,14 @@ public int getBytesSize(SeaTunnelRowType rowType) {
         return size;
     }
 
+    public boolean isBinaryFormat() {
+        return getOptions().getOrDefault("is_binary_format", false).equals(true);
+    }
+
+    public boolean isComplete() {
+        return getOptions().getOrDefault("is_complete", false).equals(true);
+    }
```
The option value belong binary format, so I think we should not put the method into `SeaTunnelRow`.

Hisoka-X commented on seatunnel-connectors-v2/connector-file/connector-file-base/src/main/java/org/apache/seatunnel/connectors/seatunnel/file/source/reader/BinaryReadStrategy.java:
```diff
@@ -49,6 +49,9 @@ public class BinaryReadStrategy extends AbstractReadStrategy {
     private boolean completeFileMode =
             FileBaseSourceOptions.BINARY_COMPLETE_FILE_MODE.defaultValue();
 
+    private final String IS_COMPLETE = "is_complete";
+    private final String IS_BINARY_FORMAT = "is_binary_format";
+
```
Move it into `CommonOptions`

xiaochen-zhou committed:
update code

Hisoka-X commented on seatunnel-connectors-v2/connector-file/connector-file-base/src/test/java/org/apache/seatunnel/connectors/seatunnel/file/reader/BinaryReadStrategyTest.java:
```diff
@@ -102,7 +104,7 @@ public void testBinaryReadWithCustomChunkSize() throws IOException {
         List<SeaTunnelRow> rows = collector.getRows();
         Assertions.assertEquals(
-                3, rows.size(), "Should have 3 chunks for 1500 bytes with 512 chunk size");
+                3 + 1, rows.size(), "Should have 3 chunks for 1500 bytes with 512 chunk size");
```
```suggestion
                3 + 1, rows.size(), "Should have 4 chunks for 1500 bytes with 512 chunk size");
```

Hisoka-X commented on seatunnel-connectors-v2/connector-file/connector-file-base/src/test/java/org/apache/seatunnel/connectors/seatunnel/file/reader/BinaryReadStrategyTest.java:
```diff
@@ -70,7 +70,9 @@ public void testBinaryReadWithDefaultChunkSize() throws IOException {
-                2, rows.size(), "Should have 2 chunks for 2048 bytes with default 1024 chunk size");
+                2 + 1,
+                rows.size(),
+                "Should have 2 chunks for 2048 bytes with default 1024 chunk size");
```
```suggestion
                "Should have 3 chunks for 2048 bytes with default 1024 chunk size");
```

Hisoka-X commented on seatunnel-connectors-v2/connector-file/connector-file-base/src/test/java/org/apache/seatunnel/connectors/seatunnel/file/reader/BinaryReadStrategyTest.java:
```diff
@@ -128,7 +130,7 @@ public void testBinaryReadCompleteFileMode() throws IOException {
 
         List<SeaTunnelRow> rows = collector.getRows();
-        Assertions.assertEquals(1, rows.size(), "Should have 1 row in complete file mode");
+        Assertions.assertEquals(1 + 1, rows.size(), "Should have 1 row in complete file mode");
```
```suggestion
        Assertions.assertEquals(1 + 1, rows.size(), "Should have 2 row in complete file mode");
```

Hisoka-X commented on docs/en/transform-v2/embedding.md:
````diff
@@ -263,6 +319,165 @@ sink {
 }
 ```
 
+### Multimodal Embedding (Volcengine Doubao)
````
Could you add an example for binary file mode?

xiaochen-zhou replied:
> Could you add an example for binary file mode?

OK.

xiaochen-zhou replied:
> Could you add an example for binary file mode?

Done.

xiaochen-zhou committed:
update doc and test

update doc

Hisoka-X commented:
Overall LGTM. Waiting test case passes.

xiaochen-zhou committed:
update code

xiaochen-zhou commented:
> Overall LGTM. Waiting test case passes.

Done.

Hisoka-X reviewed (approved):
LGTM. Thanks @xiaochen-zhou
